File: src/ChpokkWeb/SystemFiles/Templates/ProjectTemplates/VisualBasic/Silverlight/1041/BusinessApplication/Helpers/DataFieldExtensions.vb
Imports System
Imports System.Runtime.CompilerServices
Imports System.Windows
Imports System.Windows.Controls
Imports System.Windows.Data

Public Module DataFieldExtensions
    ''' <summary>
    ''' <see cref="DataField" /> の <see cref="TextBox" /> コントロールを別のコントロールに置換し、バインディングを更新します。
    ''' </summary>
    ''' <param name="field"><see cref="TextBox"/> が置換される <see cref="DataField"/> です。</param>
    ''' <param name="newControl"><see cref="DataField.Content" /> として設定する新しいコントロールです。</param>
    ''' <param name="dataBindingProperty">データ バインディングに使用されるコントロールのプロパティです。</param>        
    ''' <param name="bindingSetupFunction">
    ''' <paramref name="newControl"/> に適用される前に新しく生成されたバインディングでパラメーターの変更に使用できる、オプションの <c>Action</c> です。
    ''' </param>
    <Extension()> _
    Public Sub ReplaceTextBox(ByVal field As DataField, ByVal newControl As FrameworkElement, ByVal dataBindingProperty As DependencyProperty, _
                              Optional ByVal bindingSetupFunction As Action(Of Binding) = Nothing)
        If field Is Nothing Then
            Throw New ArgumentNullException("field")
        End If

        If newControl Is Nothing Then
            Throw New ArgumentNullException("newControl")
        End If

        ' 既存のバインディングをコピーして新しいバインディングを構築し、ユーザーが指定した関数 (ある場合) に渡します
        Dim newBinding As Binding = field.Content.GetBindingExpression(TextBox.TextProperty).ParentBinding.CreateCopy()

        If bindingSetupFunction IsNot Nothing Then
            bindingSetupFunction(newBinding)
        End If

        ' 置換フィールド
        newControl.SetBinding(dataBindingProperty, newBinding)
        field.Content = newControl
    End Sub

    ''' <summary>
    ''' 別の <see cref="Binding"/> オブジェクトからすべてのプロパティをコピーして、新しい <see cref="Binding"/> オブジェクトを作成します。
    ''' </summary>
    ''' <param name="binding">プロパティ値のコピー元の <see cref="Binding"/> です</param>
    ''' <returns>新しい <see cref="Binding"/> オブジェクトです。</returns>
    <Extension()> _
    Public Function CreateCopy(ByVal binding As Binding) As binding
        If binding Is Nothing Then
            Throw New ArgumentNullException("binding")
        End If

        Dim newBinding As New Binding()

        newBinding.BindsDirectlyToSource = binding.BindsDirectlyToSource
        newBinding.Converter = binding.Converter
        newBinding.ConverterParameter = binding.ConverterParameter
        newBinding.ConverterCulture = binding.ConverterCulture
        newBinding.Mode = binding.Mode
        newBinding.NotifyOnValidationError = binding.NotifyOnValidationError
        newBinding.Path = binding.Path
        newBinding.UpdateSourceTrigger = binding.UpdateSourceTrigger
        newBinding.ValidatesOnExceptions = binding.ValidatesOnExceptions

        If binding.ElementName IsNot Nothing Then
            newBinding.ElementName = binding.ElementName
        ElseIf binding.RelativeSource IsNot Nothing Then
            newBinding.RelativeSource = binding.RelativeSource
        Else
            newBinding.Source = binding.Source
        End If

        Return newBinding
    End Function
End Module

File: src/ChpokkWeb/SystemFiles/Templates/ProjectTemplates/CSharp/Silverlight/1036/BusinessApplication/Helpers/DataFieldExtensions.cs
namespace $safeprojectname$
{
    using System;
    using System.Windows;
    using System.Windows.Controls;
    using System.Windows.Data;

    /// <summary>
    ///     Fournit des méthodes d'extension permettant d'exécuter des opérations sur un <see cref="DataField"/>.
    /// </summary>
    public static class DataFieldExtensions
    {
        /// <summary>
        /// Remplace le contrôle <see cref="TextBox" /> d'un <see cref="DataField" /> par un autre contrôle et met à jour les liaisons.
        /// </summary>
        /// <param name="field"><see cref="DataField"/> dont <see cref="TextBox"/> sera remplacé.</param>
        /// <param name="newControl">Nouveau contrôle que vous allez définir en tant que <see cref="DataField.Content" />.</param>
        /// <param name="dataBindingProperty">Propriété du contrôle qui sera utilisée pour la liaison de données.</param>        
        public static void ReplaceTextBox(this DataField field, FrameworkElement newControl, DependencyProperty dataBindingProperty)
        {
            field.ReplaceTextBox(newControl, dataBindingProperty, null);
        }

        /// <summary>
        /// Remplace le contrôle <see cref="TextBox" /> d'un <see cref="DataField" /> par un autre contrôle et met à jour les liaisons.
        /// </summary>
        /// <param name="field"><see cref="DataField"/> dont <see cref="TextBox"/> sera remplacé.</param>
        /// <param name="newControl">Nouveau contrôle que vous allez définir en tant que <see cref="DataField.Content" />.</param>
        /// <param name="dataBindingProperty">Propriété du contrôle qui sera utilisée pour la liaison de données.</param>        
        /// <param name="bindingSetupFunction">
        ///  <see cref="Action"/> optionnelle que vous pouvez utiliser pour modifier des paramètres sur la liaison qui vient d'être créée avant qu'elle soit appliquée à <paramref name="newControl"/>
        /// </param>
        public static void ReplaceTextBox(this DataField field, FrameworkElement newControl, DependencyProperty dataBindingProperty, Action<Binding> bindingSetupFunction)
        {
            if (field == null)
            {
                throw new ArgumentNullException("field");
            }

            if (newControl == null)
            {
                throw new ArgumentNullException("newControl");
            }

            // Construire une nouvelle liaison en en copiant une existante, et en l'envoyant à bindingSetupFunction pour que l'appelant puissent apporter les modifications qu'il souhaite.
            Binding newBinding = field.Content.GetBindingExpression(TextBox.TextProperty).ParentBinding.CreateCopy();

            if (bindingSetupFunction != null)
            {
                bindingSetupFunction(newBinding);
            }

            // Remplacer le champ
            newControl.SetBinding(dataBindingProperty, newBinding);
            field.Content = newControl;
        }

        /// <summary>
        /// Crée un nouvel objet <see cref="Binding"/> en copiant toutes les propriétés d'un autre objet <see cref="Binding"/>.
        /// </summary>
        /// <param name="binding"><see cref="Binding"/> à partir duquel les valeurs de propriétés seront copiées</param>
        /// <returns>Nouvel objet <see cref="Binding"/>.</returns>
        private static Binding CreateCopy(this Binding binding)
        {
            if (binding == null)
            {
                throw new ArgumentNullException("binding");
            }

            Binding newBinding = new Binding()
            {
                BindsDirectlyToSource = binding.BindsDirectlyToSource,
                Converter = binding.Converter,
                ConverterParameter = binding.ConverterParameter,
                ConverterCulture = binding.ConverterCulture,
                Mode = binding.Mode,
                NotifyOnValidationError = binding.NotifyOnValidationError,
                Path = binding.Path,
                UpdateSourceTrigger = binding.UpdateSourceTrigger,
                ValidatesOnExceptions = binding.ValidatesOnExceptions
            };

            if (binding.ElementName != null)
            {
                newBinding.ElementName = binding.ElementName;
            }
            else if (binding.RelativeSource != null)
            {
                newBinding.RelativeSource = binding.RelativeSource;
            }
            else
            {
                newBinding.Source = binding.Source;
            }

            return newBinding;
        }
    }
}
